### ISSUE TITLE
Enhancement: [explicit-function-return-type] support patterns in `allowedNames`

### ISSUE THREAD

**Issue by dsokal** [typescript-eslint/typescript-eslint#7876]
### Before You File a Proposal Please Confirm You Have Done The Following...

- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22enhancement%3A+plugin+rule+option%22) and found none that match my proposal.
- [X] I have searched the [current rule list](https://typescript-eslint.io/rules/#supported-rules) and found no rules that match my proposal.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### My proposal is suitable for this project

- [X] I believe my proposal would be useful to the broader TypeScript community (meaning it is not a niche proposal).

### Link to the rule's documentation

https://typescript-eslint.io/rules/explicit-function-return-type/

### Description

I propose that the `explicit-function-return-type` rule should let me specify a regexp in `allowedNames` so I can ignore the rule for functions with names matching the regexp.
My use case is that I have functions returning complex objects with `as const`. I don't want to type the return types of those functions as it's usually pointless and too time-consuming for me. Those functions always end with a particular suffix. I'd love to ignore the rule at the eslintrc level. 
If allowing for regular expressions in `allowedNames` sounds too risky, e.g. someone could add `.*` as one of the allowed names and so it'd disable the rule for all functions - my another idea is to allow for another item type for `allowedNames`, e.g. `{ regexp: 'pa(t){3,5}ern' }`. This would look sth like: `{ allowedNames: ['string', { regexp: 'pattern' }]}`

### Fail

```typescript
// { allowedNames: ['.*pa(t){3,5}ern$'] }

function notInAllowedNames() {
  return 123;
}

const foo = function xyzPatttern() {
  return {
    foo: 'bar',
  } as const;
};
```

### Pass

```typescript
// { allowedNames: ['.*pa(t){3,5}ern$'] }

function notInAllowedNames(): number {
  return 123;
}

const foo = function xyzPatttern() {
  return {
    foo: 'bar',
  } as const;
};
```

### Additional Info

I've already put up a PR - https://github.com/typescript-eslint/typescript-eslint/pull/7869

**Comment by bradzacher** [comment]
Duplicate of https://github.com/typescript-eslint/typescript-eslint/issues/7053
